core/src/main/java/io/questdb/cairo/O3MemoryPressureRegulatorImpl.java

```diff
@@ -0,0 +1,160 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo;
+
+import io.questdb.cairo.wal.seq.SeqTxnTracker;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.Rnd;
+import io.questdb.std.datetime.microtime.MicrosecondClock;
+import org.jetbrains.annotations.TestOnly;
+
+/**
+ * This class is responsible for regulating the memory pressure in the O3 WAL apply system.
+ * <p>
+ * It is used to control the parallelism of the O3 merge process and to introduce backoff
+ * when the system is under memory pressure.
+ * <p>
+ * This class is NOT thread-safe.
+ */
+public final class O3MemoryPressureRegulatorImpl implements O3MemoryPressureRegulator {
+    private static final Log LOG = LogFactory.getLog(O3MemoryPressureRegulatorImpl.class);
+
+    private static final int MAX_LEVEL = 10;
+    private static final int MIN_LEVEL = 0;
+    private static final int PARALLELISM_THROTTLING_LEVEL = 5; // when exceeded we start introducing back off
```

Comment:
Usually, everything is configurable, this should not be an exception

Reply:
@ideoma there are still hard-coded values. Do you want to externalize them? 
I was thinking about a switch to disable it entirely if it goes bananas. But I would rather not to expose any find tuning parameters since they are tightly coupled to a specific regulator strategy and it can and likely will change. 
wdyt?

Reply:
The current ones make even less sense to be configurable -- upper bound on backoff wait time, and likelihood of increasing parallelism.